src/main/java/com/google/devtools/build/lib/actions/ExecutionRequirements.java

```diff
@@ -105,6 +105,96 @@ public String parseIfMatches(String tag) throws ValidationException {
     }
   }
 
+  /** An execution requirement that is split into a key and a parseable value. */
+  @AutoValue
+  public abstract static class TaggedRequirement<ValueType> {
+    /**
+     * Thrown when a the {@link #validator()} method returns an error.
+     */
+    public static class ValidationException extends Exception {
+      private final String tagValue;
+
+      /**
+       * Creates a new {@link ValidationException}.
+       *
+       * @param tagValue the erroneous value that was parsed from the tag.
+       * @param errorMsg an error message that tells the user what's wrong with the value.
+       */
+      public ValidationException(String tagValue, String errorMsg) {
+        super(errorMsg);
+        this.tagValue = tagValue;
+      }
+
+      /**
+       * Returns the erroneous value of the parsed tag.
+       *
+       * <p>Useful to put in error messages shown to the user.
+       */
+      public String getTagValue() {
+        return tagValue;
+      }
+    }
+
+    /**
+     * Create a new parseable execution requirement definition.
+     *
+     * @param tagName a human readable name of the tag
+     * @param validator a Function that will be used to validate the value of the tag. It should
+     *     return null if the value is fine to use or a human-friendly error message describing why
+     *     the value is not valid.
+     * @param parser a Function that will transform the value of the tag to {@link ValueType}.
+     */
+    static <ValueType> TaggedRequirement<ValueType> create(
+            String tagName,
+            Function<String, String> validator,
+            Function<String, ValueType> parser) {
+      return new AutoValue_ExecutionRequirements_TaggedRequirement<ValueType>(
+          tagName, validator, parser);
+    }
+
+    public abstract String tagName();
+
+    public abstract Function<String, String> validator();
+
+    public abstract Function<String, ValueType> parser();
+
+    /**
+     * Returns the parsed value from a tag's value.
+     *
+     * @throws ValidationException if the value parsed out of the tag doesn't pass the validator.
+     */
+    public ValueType parse(String value) throws ValidationException {
+      String errorMsg = validator().apply(value);
+      if (errorMsg != null) {
+        throw new ValidationException(value, errorMsg);
+      }
+      return parser().apply(value);
+    }
+  }
+
+  /** Integer validator to use with {@link ParseableRequirement} and {@link TaggedRequirement}. */
+  private static String IntegerValidator(String s) {
+    Preconditions.checkNotNull(s);
```

Comment:
When would we expect this to be null? I think never, so it's fair to assume it's not null and let the code below crash if it is.

Reply:
As a traveler in a foreign codebase, I'm totally okay with removing it, but that check predates me, fwiw. (See da21ba7a48e.)